### PR TITLE
feat: add InAnyOrder matcher

### PR DESCRIPTION
**Pull request by vvkh**
Draft implementation for #545

### CODE REVIEW

File: gomock/matchers.go [hunk 1]
```diff
@@ -207,6 +207,66 @@ func (m lenMatcher) String() string {
...
+	given, ok := m.prepareValue(x)
+	if !ok {
+		return false
+	}
+	wanted, ok := m.prepareValue(m.x)
+	if !ok {
+		return false
+	}
+
+	usedFromGiven := make([]bool, given.Len())
```
Comment (codyoss) [inline gomock/matchers.go:224]: Maybe we should fail fast here if lens don't match, they should be exact right?

File: gomock/matchers.go [hunk 2]
```diff
@@ -207,6 +207,66 @@ func (m lenMatcher) String() string {
...
+		}
+	}
+
+	return extraInGiven == 0 && missingFromWanted == 0
+}
+
+func (m inAnyOrderMatcher) prepareValue(x interface{}) (reflect.Value, bool) {
+	xValue := reflect.ValueOf(x)
+	switch xValue.Kind() {
+	case reflect.Slice, reflect.Array, reflect.String:
```
Comment (codyoss) [inline gomock/matchers.go:259]: I am not sure it makes sense here to also allow strings. I think we should just stick to arrays and slices for now.

File: gomock/matchers_test.go [hunk 3]
```diff
@@ -142,3 +142,176 @@ func TestAssignableToTypeOfMatcher(t *testing.T) {
...
+
+func TestInAnyOrder(t *testing.T) {
+	tests := []struct {
+		name      string
+		wanted    interface{}
+		given     interface{}
+		wantMatch bool
+	}{
+		{
+			name:      "match_for_equal_slices",
```
Comment (codyoss) [inline gomock/matchers_test.go:154]: nit: can use spaces instead of underscores for test names.

**Review by codyoss — COMMENTED** [review]
Just a couple of small things, thanks for the contribution!

File: gomock/matchers.go (hunk 1 above)
Comment (vvkh) [reply gomock/matchers.go:228]: Good point, added early return for this case.

File: gomock/matchers.go (hunk 2 above)
Comment (vvkh) [reply gomock/matchers.go]: Ok, we can add it later if someone actually needs this. 👌

File: gomock/matchers_test.go (hunk 3 above)
Comment (vvkh) [reply gomock/matchers_test.go:154]: Sure, it's just nice to be able to copy name of a test with underscores directly from gotest log as is and search for it in codebase. :)

Comment (codyoss) [reply gomock/matchers_test.go:154]: Looks like this was not resolved.

**Review by codyoss — CHANGES_REQUESTED** [review]
Just one minor thing, other than that looks good to me!

**Review by codyoss — APPROVED** [review]
LGTM, thanks for the contribution!

**Pull request merged**